tools/bumper/git.go

```diff
@@ -0,0 +1,175 @@
+	if err != nil {
+		return errors.Wrapf(err, "Failed to get user from github client API")
+	}
+	return nil
+}
+
+func (mgr *gitHubManager) getCurrentReleaseTag(component component, repoDir string) (string, error) {
+	repo := getComponentNameFromUrl(component.Url)
+	owner := getComponentOwnerFromUrl(component.Url)
+	logger.Printf("Getting current tag in repo %s sha %s", repo, component.Commit)
+
+	tagRefs, _, err := mgr.client.Git.ListMatchingRefs(mgr.ctx, owner, repo, &github.ReferenceListOptions{Ref: "refs/tags"})
+	if err != nil {
+		return "", errors.Wrap(err, "Failed to get release tag refs from github client API")
+	}
+
+	// look for the tag that belongs to the chosen commit, going over from last tag
+	for i := len(tagRefs) - 1; i >= 0; i-- {
+		tag := tagRefs[i]
+		tagName := strings.Replace(tag.GetRef(), "refs/tags/", "", 1)
+		commitShaOfTag, err := getTagCommitSha(tagName, repoDir)
+		if err != nil {
+			return "", errors.Wrap(err, "Failed to get commit-sha from tag name")
+		}
+
+		if component.Commit == commitShaOfTag {
+			return tagName, nil
+		}
+	}
+
+	//if not found, then return the vtag of the current commit
+	return getVirtualTag(component.Commit, repoDir)
+}
+
+func (mgr *gitHubManager) getUpdatedReleaseInfo(component component, repoDir string) (string, string, error) {
+	repo := getComponentNameFromUrl(component.Url)
+	owner := getComponentOwnerFromUrl(component.Url)
+	switch component.Updatepolicy {
+	case updatePolicyTagged:
+		logger.Printf("update policy %s will updated to the latest tagged commit in the referenced branch", updatePolicyTagged)
+		return mgr.getLatestTaggedFromBranch(repo, owner, component.Branch, repoDir)
+	case updatePolicyLatest:
+		logger.Printf("update policy %s will update to the latest HEAD in the referenced branch", updatePolicyTagged)
+		return mgr.getLatestFromBranch(repo, owner, component.Branch, repoDir)
+	case updatePolicyStatic:
+		logger.Printf("update policy %s will disable auto update", updatePolicyStatic)
+		return component.Metadata, component.Commit, nil
+	default:
+		return "", "", fmt.Errorf("Error: Update strategy %s not supported", component.Updatepolicy)
+	}
+}
+
+func (mgr *gitHubManager) getLatestTaggedFromBranch(repo, owner, branch, repoDir string) (string, string, error) {
+	logger.Printf("Getting latest tagged from branch %s in repo %s", branch, repo)
+	tagRefs, _, err := mgr.client.Git.ListMatchingRefs(mgr.ctx, owner, repo, &github.ReferenceListOptions{Ref: "refs/tags"})
+	if err != nil {
+		return "", "", errors.Wrap(err, "Failed to get release tag refs from github client API")
+	}
+
+	branchCommits, _, err := mgr.client.Repositories.ListCommits(mgr.ctx, owner, repo, &github.CommitsListOptions{SHA: branch})
+	if err != nil {
+		return "", "", errors.Wrap(err, "Failed to get release tag refs from github client API")
+	}
+
+	// look for the first tag that belongs to the chosen branch, going over from last tag
+	for i := len(tagRefs) - 1; i >= 0; i-- {
```

Comment:
Can't you iterate over the range?

Reply:
I can, but the latest tags are always last. 
So in order to find it faster I run from the end (regular golang iteration doesn't have a "reverse" run option)...

Reply:
Why don't we just take the last item? Is one of the steps expected to fail?

Reply:
the last one may not belong to the branch specified. 
We want the latest tag that belongs to this branch, and no list (tag refs/branch commits) holds them both, so we need to so a cross search to find it.